src/mlpack/methods/ann/dists/bernoulli_distribution_impl.hpp

```diff
@@ -0,0 +1,85 @@
+/**
+ * @file bernoulli_distribution_impl.hpp
+ * @author Atharva Khandait
+ *
+ * Implementation of the Bernoulli distribution class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_DISTRIBUTIONS_BERNOULLI_DISTRIBUTION_IMPL_HPP
+#define MLPACK_METHODS_ANN_DISTRIBUTIONS_BERNOULLI_DISTRIBUTION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "bernoulli_distribution.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename DataType>
+BernoulliDistribution<DataType>::BernoulliDistribution() :
+    applyLogistic(true),
+    eps(1e-10)
+{
+  // Nothing to do here.
+}
+
+template<typename DataType>
+BernoulliDistribution<DataType>::BernoulliDistribution(
+    const DataType&& param,
+    const bool applyLogistic,
+    const double eps) :
+    logits(param),
+    applyLogistic(applyLogistic),
+    eps(eps)
+{
+  if (applyLogistic)
+    LogisticFunction::Fn(logits, probability);
+  else
+  {
+    probability = arma::mat(logits.memptr(), logits.n_rows,
+        logits.n_cols, false, false);
+  }
+}
+
+template<typename DataType>
+DataType BernoulliDistribution<DataType>::Sample() const
+{
+  DataType sample;
+  sample = arma::conv_to<DataType>::from(arma::randu<DataType>
+      (probability.n_rows, probability.n_cols) <= probability);
```

Comment:
If this turns out to be a bottleneck, it might be possible to optimize by looping directly over the elements of `sample`.  (i.e. the conversion to `DataType` from `arma::uword` won't be optimized out at compile time, I don't think)

Reply:
Oh, so do you mean something like this:
```cpp
Datatype sample = Datatype(size(probability));
for (int i = 0; i < probability.n_elem; i++)
{
  sample(i) = (int)randu(i)  < probability(i);
}
```
Also, can you please explain what exactly you mean when you say it won't be optimized at compile time? Can you point me to some source where I can read about it a bit?

Reply:
@rcurtin Sorry for reminding you, I am sure you would have gotten back on this, but I need to get this merged soon. Can you please answer the above doubt?

Reply:
Hey Atharva,

Sorry, it was a really busy end-of-week and weekend and I didn't get to this.  I think the code is fine but you should cast to `Datatype` not `int`.

For the compile-time optimizations, Armadillo is a library that uses delayed evaluation.  So, basically, if I write 'matrix + matrix', the type returned is not a matrix, but instead a placeholder type that indicates that two matrices need to be added.

Section 8 of this paper has some more details:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.739.504&rep=rep1&type=pdf

Also, here are the slides from a talk I gave that has more details:
http://conradsanderson.id.au/pdfs/sanderson_curtin_armadillo_pasc_2017.pdf

I hope this helps, let me know if I can clarify anything.  And sorry for the slow response. :)

Reply:
Thanks for the response!